src/tinyformat.h

```diff
@@ -179,13 +179,19 @@ namespace tfm = tinyformat;
 
 namespace tinyformat {
 
+// Similar to std::runtime_format from C++26.
+struct RuntimeFormat {
+    explicit RuntimeFormat(std::string str) : fmt{std::move(str)} {}
+    std::string fmt;
+};
+
 // Added for Bitcoin Core. Wrapper for checking format strings at compile time.
 // Unlike ConstevalFormatString this supports std::string for runtime string
 // formatting without compile time checks.
 template <unsigned num_params>
 struct FormatStringCheck {
     consteval FormatStringCheck(const char* str) : fmt{util::ConstevalFormatString<num_params>{str}.fmt} {}
-    FormatStringCheck(const std::string& str) : fmt{str.c_str()} {}
+    FormatStringCheck(const RuntimeFormat& run) : fmt{run.fmt.c_str()} {}
```

Comment:
Just realized there's some risk of dangling pointers.
```suggestion
    FormatStringCheck(const RuntimeFormat& run LIFETIMEBOUND) : fmt{run.fmt.c_str()} {}
```

Also, if you're going to rename this one `str`->`run`, maybe call the other one `comp` or something?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1875730141

Delete seems good, but I think maybe LIFETIMEBOUND could still be helpful because `fmt` member here is being assigned the result of `run.fmt.c_str()`, so the `RuntimeFormat` object lifetime needs to be at least as long as `FormatStringCheck` object lifetime, and maybe LIFETIMEBOUND could help enforce this? I don't know all of the details about how LIFETIMEBOUND works though, so this is just a question.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1878803831

I'm still not sure I see any downsides to adding LIFETIMEBOUND. It seems like a great way to document the fact that these functions require long-lived references, and to provide instant feedback for code and design bugs without needing to create a PR and wait for an ASAN job to fail.

I also don't understand why the additional temporaries and moves and elisions in fad431e548a694600a3fe212127052e5f922a622 would preferable to using references and creating the minimum number of objects with straightforward full-expression lifetimes.

fad431e548a694600a3fe212127052e5f922a622 also seems unsafe because it doesn't prevent lvalues from being passed to the FormatStringCheck constructor, and if an lvalue (or any other argument incompatible with copy elision) is passed it will result in `FormatStringCheck::fmt` being assigned a dangling pointer.

Would suggest following changes as a simpler alternative:

```diff
--- a/src/tinyformat.h
+++ b/src/tinyformat.h
@@ -141,6 +141,7 @@ namespace tfm = tinyformat;
 
 //------------------------------------------------------------------------------
 // Implementation details.
+#include <attributes.h> // Added for Bitcoin Core
 #include <algorithm>
 #include <iostream>
 #include <sstream>
@@ -181,10 +182,8 @@ namespace tinyformat {
 
 // Added for Bitcoin Core. Similar to std::runtime_format from C++26.
 struct RuntimeFormat {
-    std::string fmt; // Not a string view, because tinyformat requires a c_str
-    explicit RuntimeFormat(std::string str) : fmt{std::move(str)} {}
-    RuntimeFormat(const RuntimeFormat&) = delete;
-    RuntimeFormat& operator=(const RuntimeFormat&) = delete;
+    const std::string& fmt; // Not a string view, because tinyformat requires a c_str
+    explicit RuntimeFormat(LIFETIMEBOUND const std::string& str) : fmt{str} {}
 };
 
 // Added for Bitcoin Core. Wrapper for checking format strings at compile time.
@@ -193,7 +192,7 @@ struct RuntimeFormat {
 template <unsigned num_params>
 struct FormatStringCheck {
     consteval FormatStringCheck(const char* str) : fmt{util::ConstevalFormatString<num_params>{str}.fmt} {}
-    FormatStringCheck(const RuntimeFormat& run) : fmt{run.fmt.c_str()} {}
+    FormatStringCheck(LIFETIMEBOUND const RuntimeFormat& run) : fmt{run.fmt.c_str()} {}
     FormatStringCheck(util::ConstevalFormatString<num_params> str) : fmt{str.fmt} {}
     operator const char*() { return fmt; }
     const char* fmt;
```

Reply:
`LIFETIMEBOUND` only being supported by one compiler isn't a great reason to skip adding it. Good to let Clang do what it can, and also provide documentation to devs. Not blocking for me though.

Tried the following, but Clang 18 compiles without any warnings when adding `LIFETIMEBOUND`. 
```C++
std::optional<tinyformat::FormatStringCheck<0>> fsc;
{
    auto fmt{std::make_unique<tinyformat::RuntimeFormat>("{}")};
    fsc = {tinyformat::FormatStringCheck<0>(*fmt)};
}
```

Also tried removing the `RuntimeFormat`-ctor from `FormatStringCheck` and found that it's only being used in `tinyformat::format(util::BilingualFmt...` outside off fuzz-tests. So this should be a minor concern.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31061#discussion_r1879761314

> Apart from the downsides I mentioned,
>
> It is just one reason, among many that have been mentioned previously.

Thanks for adding LIFETIMEBOUND here. I've been pretty confused by this discussion because I haven't seen any reasons given for why it would actually be better not to use LIFETIMEBOUND, and I don't know of any reasons. I understand LIFETIMEBOUND doesn't work everywhere, but don't understand why that matters as long as it catches bugs and provides fast feedback in the places it does work, and is helpful everywhere to document references that need to be long-lived.

If LIFETIMEBOUND triggered false positive errors like clang thread safety annotations do, that could be a real downside and reason not to use it. But as far as I know it doesn't cause harms or have any negative effects like that, so I've been confused by the pushback here. Maybe the reasons not to use it are as simple as "looks ugly"  / "adds noise"  / "seems useless" which are perfectly ok (and are also reasons I don't use `const` various places). i'd be fine with keeping or dropping LIFETIMEBOUND in this PR, just wondering if I am not understanding real problems with it.

Reply:
Oh thanks for explaining that. I thought you were only disabling copies and assignment to prevent expensive copies due to not passing RuntimeFormat by reference, and that you were not passing RuntimeFormat by reference to avoid needing to think about lifetime issues.

But actually your goal is to make `RuntimeFormat run("fmt"); strprintf(run);` a compiler error, and deleting the copy and assignment operators are a way to do that. So LIFETIMEBOUND isn't harmful, but just not as good 
as a potentially safer solution fad431e548a694600a3fe212127052e5f922a622 that doesn't use LIFETIMEBOUND.

I think I can see now how fad431e548a694600a3fe212127052e5f922a622 actually is a safer solution in every case, and relying on mandatory copy elision is ok there because deleting the RuntimeFormat copy constructor should make it impossible to pass as a function argument without copy elision. This is more complicated than just using references, but does seem safer.

Reply:
>  https://godbolt.org/z/6336nGKcM (replace `0` with `1` to enable the compile-time checks)) into a compile failure on all compilers seems better than a runtime error, possibly only in CI.

With the `#if` set to `1` and lines 42-43 changed to
```C++
        check = RuntimeFormat("fmt");
```
One still gets an ASAN error instead of a compile time error.